src/next_array.rs

```diff
@@ -0,0 +1,64 @@
+use core::mem::MaybeUninit;
+use core::ptr;
+
+/// Helper struct to build up an array element by element.
+struct ArrayBuilder<T, const N: usize> {
+    arr: [MaybeUninit<T>; N], // Invariant: arr[..len] is initialized.
+    len: usize,               // Invariant: len <= N.
+}
+
+impl<T, const N: usize> ArrayBuilder<T, N> {
+    pub fn new() -> Self {
+        Self {
+            arr: [(); N].map(|_| MaybeUninit::uninit()),
+            len: 0,
+        }
+    }
+
+    pub fn push(&mut self, value: T) {
+        // We maintain the invariant here that arr[..len] is initialized.
+        // Indexing with self.len also ensures self.len < N, and thus <= N after
+        // the increment.
+        self.arr[self.len] = MaybeUninit::new(value);
+        self.len += 1;
+    }
+
+    pub fn take(&mut self) -> Option<[T; N]> {
+        if self.len == N {
+            // Take the array, resetting the length back to zero.
+            let arr = core::mem::replace(&mut self.arr, [(); N].map(|_| MaybeUninit::uninit()));
+            self.len = 0;
+
+            // SAFETY: we had len == N, so all elements in arr are initialized.
+            Some(unsafe { arr.map(|v| v.assume_init()) })
+        } else {
+            None
+        }
+    }
```

Comment:
Slightly rearranging operations here allows us to preserve `ArrayBuilder`'s invariants and avoid temporarily violating them:
```suggestion
    /// Attempts to convert `self` into an array `[T; N]`.
    ///
    /// If `self.len() < N`, this returns `None`.
    pub fn take(&mut self) -> Option<[T; N]> {
        let len = self.len;
        if len == N {
            // SAFETY: Decreasing the value of `self.len` cannot violate its
            // safety invariant.
            self.len = 0;

            // SAFETY: Since `self.len` is 0, `self.arr` may safely contain
            // uninitialized elements.
            let arr = mem::replace(&mut self.arr, [(); N].map(|_| MaybeUninit::uninit()));

            let arr = arr.map(|elt| {
                // SAFETY: We know that all elements of `arr` are valid because
                // we checked that `len == N`.
                unsafe { elt.assume_init() }
            });

            Some(arr)
        } else {
            None
        }
    }
```

Reply:
I will reorder those operations, sure.